src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -24,88 +24,123 @@
 
 package com.fulcrumgenomics.util
 
+import com.fulcrumgenomics.util.ReadStructure.{SubReadWithQuals, SubReadWithoutQuals}
+import com.fulcrumgenomics.FgBioDef._
+
 import scala.collection.immutable
-import scala.reflect.ClassTag
+import scala.collection.mutable.ArrayBuffer
 
-// Future improvement: I think we really need to support having a final segment of indeterminate length. Probably it
-// would be good to have support for * and + to indicate zero or more and one or more bases.
+/**
+  * Companion object for ReadStructure that provides factory methods.
+  */
 object ReadStructure {
+  /** A character that can be put in place of a number in a read structure to mean "0 or more bases". */
+  val AnyLengthChar: Char = '+'
+
+  /** Can be subtracted from any digit character to get it's integer value. */
+  private val DigitOffset = '0'.toInt
+
+  /** Contains the bases and optionally base qualities that correspond to the given read segment. */
+  sealed trait SubRead {
+    def bases: String
+    def segment: ReadSegment
+    def kind: SegmentType = this.segment.kind
```

Comment:
My only concern would be that we would want some information from `ReadSegment` (i.e. `length` or `offset`) that we would lose by getting rid of it...
![](http://www.reactiongifs.com/r/nsr.gif)